synapse/storage/databases/main/room.py

```diff
@@ -540,23 +553,31 @@ async def get_rooms_paginate(
             # Flip the boolean
             order_by_asc = not order_by_asc
 
+        if where_statement:
+            info_where_statement = where_statement + " AND events.type='m.room.create'"
+        else:
+            info_where_statement = "WHERE events.type='m.room.create'"
+
         # Create one query for getting the limited number of events that the user asked
         # for, and another query for getting the total number of events that could be
         # returned. Thus allowing us to see if there are more events to paginate through
         info_sql = """
             SELECT state.room_id, state.name, state.canonical_alias, curr.joined_members,
               curr.local_users_in_room, rooms.room_version, rooms.creator,
               state.encryption, state.is_federatable, rooms.is_public, state.join_rules,
-              state.guest_access, state.history_visibility, curr.current_state_events
+              state.guest_access, state.history_visibility, curr.current_state_events,
+              json.json
             FROM room_stats_state state
             INNER JOIN room_stats_current curr USING (room_id)
             INNER JOIN rooms USING (room_id)
+            INNER JOIN event_json json USING (room_id)
+            INNER JOIN state_events events USING (event_id)
```

Comment:
This is a comparison of costs on my small database for `get_rooms_paginate`.
It seems to be a big growth. I think we need some discussion to this: https://github.com/matrix-org/synapse/issues/11904#issuecomment-1032430625

```sql
explain
SELECT state.room_id, state.name, state.canonical_alias, curr.joined_members,
  curr.local_users_in_room, rooms.room_version, rooms.creator,
  state.encryption, state.is_federatable, rooms.is_public, state.join_rules,
  state.guest_access, state.history_visibility, curr.current_state_events
FROM room_stats_state state
INNER JOIN room_stats_current curr USING (room_id)
INNER JOIN rooms USING (room_id);
                                      QUERY PLAN
--------------------------------------------------------------------------------------
 Hash Join  (cost=11.31..16.20 rows=112 width=218)
   Hash Cond: (state.room_id = curr.room_id)
   ->  Hash Join  (cost=5.57..10.14 rows=114 width=262)
         Hash Cond: (state.room_id = rooms.room_id)
         ->  Seq Scan on room_stats_state state  (cost=0.00..4.24 rows=124 width=159)
         ->  Hash  (cost=4.14..4.14 rows=114 width=103)
               ->  Seq Scan on rooms  (cost=0.00..4.14 rows=114 width=103)
   ->  Hash  (cost=4.22..4.22 rows=122 width=67)
         ->  Seq Scan on room_stats_current curr  (cost=0.00..4.22 rows=122 width=67)
```

```sql
explain
SELECT state.room_id, state.name, state.canonical_alias, curr.joined_members,
  curr.local_users_in_room, rooms.room_version, rooms.creator,
  state.encryption, state.is_federatable, rooms.is_public, state.join_rules,
  state.guest_access, state.history_visibility, curr.current_state_events,
  json.json
FROM room_stats_state state
INNER JOIN room_stats_current curr USING (room_id)
INNER JOIN rooms USING (room_id)
INNER JOIN event_json json USING (room_id)
INNER JOIN state_events events USING (event_id);
                                              QUERY PLAN
-------------------------------------------------------------------------------------------------------
 Hash Join  (cost=639.49..5749.86 rows=8125 width=1289)
   Hash Cond: (state.room_id = curr.room_id)
   ->  Hash Join  (cost=633.74..5721.66 rows=8270 width=1389)
         Hash Cond: (state.room_id = rooms.room_id)
         ->  Hash Join  (cost=628.18..5691.67 rows=8995 width=1286)
               Hash Cond: (json.room_id = state.room_id)
               ->  Hash Join  (cost=622.39..5661.45 rows=8995 width=1127)
                     Hash Cond: (json.event_id = events.event_id)
                     ->  Seq Scan on event_json json  (cost=0.00..4982.49 rows=21549 width=1172)
                     ->  Hash  (cost=509.95..509.95 rows=8995 width=45)
                           ->  Seq Scan on state_events events  (cost=0.00..509.95 rows=8995 width=45)
               ->  Hash  (cost=4.24..4.24 rows=124 width=159)
                     ->  Seq Scan on room_stats_state state  (cost=0.00..4.24 rows=124 width=159)
         ->  Hash  (cost=4.14..4.14 rows=114 width=103)
               ->  Seq Scan on rooms  (cost=0.00..4.14 rows=114 width=103)
   ->  Hash  (cost=4.22..4.22 rows=122 width=67)
         ->  Seq Scan on room_stats_current curr  (cost=0.00..4.22 rows=122 width=67)
```

Reply:
@dklimpel I guess, one should more efficiently store the room `type` when the room is created. However, I did not dare such a big change.

Reply:
Also, this would not be downwards compatible

Reply:
IMO it is possible to do this downwards compatible. But then it is not a "good first issue".
You need:
- a database update to extend the table with a new column
- update the function, which writes new rooms to database
- a background job to update database for all existing / old rooms

Reply:
Ok, maybe I will do so in a month or so. I am a student and currently we have exam phase.